devito/at_controller.py

```diff
@@ -0,0 +1,245 @@
+
+        if not self.op.propagator.cache_blocking:
+            raise EnvironmentError("Invalid parameters. Cache_blocking is set to False")
+
+        logger.info("Starting auto tuning of block sizes using brute force")
+
+        self.op.auto_tune = True
+        self.op.propagator.auto_tune = True
+
+        f, args = self.op.apply(auto_tune=True)
+
+        if maximum - minimum > 2:  # No point to estimate if diff is below 2.
+            self._estimate_b_run_time(f, args, minimum, maximum)
+
+        times = []  # list where times and block sizes will be kept
+        block_list = []  # used to make sure we do not test the same block sizes
+        org_block_sizes = self.op.propagator.block_sizes
+        blocks = list(self.op.propagator.block_sizes)
+
+        for x in range(minimum, maximum):
+            blocks[0] = x if org_block_sizes[0] else None
+
+            if len(blocks) > 1:
+                for y in range(minimum, maximum):
+                    blocks[1] = y if org_block_sizes[1] else None
+
+                    if len(blocks) > 2:
+                        for z in range(minimum, maximum):
+                            blocks[2] = z if org_block_sizes[2] else None
+                            block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+                    else:
+                        block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+            else:
+                block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+
+        # runs function for each block_size
+        for block in block_list:
+            times.append((block, self._time_blocks(f, args, block)))
+
+        logger.info("Auto tuning using brute force complete")
+
+        times = sorted(times, key=lambda element: element[1])  # sorts the list of tuples based on time
+        self._write_b_report(times)  # writes the report
+
+    def _estimate_b_run_time(self, f, args, minimum, maximum):
+        """
+        Estimates run time for auto tuning
+        :param f: compiled function that we are auto tuning
+        :param args: arguments for that function
+        :param minimum: int - minimum tune range
+        :param maximum: int - maximum tune range
+        """
+        timing_run = 0  # estimating run time
+        for i in range(0, 5):
+            logger.info('Estimating auto-tuning runtime...sample %d/5' % (i + 1))
+
+            blocks = []
+            for block in self.op.propagator.block_sizes:
+                blocks = blocks + [random.randrange(minimum, maximum)] if block else blocks + [None]
+
+            timing_run += self._time_blocks(f, args, blocks)
+
+        logger.info("Estimated runtime: %f minutes." % float(
+            timing_run / 5 * ((maximum - minimum) * (maximum - minimum)) / 600))
+
+    def _time_blocks(self, f, args, block_sizes):
```

Comment:
I think this Python-level timing becomes obsolete with PR #80, since you can pull this information straight out of the Propagator now, and the timings are created actually come from the kernel itself.